### ISSUE TITLE
Very slow performance

### ISSUE THREAD

**Issue by zeridon** [GjjvdBurg/HugoPhotoSwipe#10]
I decided to try the galery generation and the performance is let's say less than slowish.

The setup is as follows:

 * Python - 2.7.3
 * HugoPhotoSwipe - 0.0.12 (installed from pip)
 * hugophotoswipe.yml (attached). It is default with just markdown_dir and output_dir set.
[hugophotoswipe.txt](https://github.com/GjjvdBurg/HugoPhotoSwipe/files/867208/hugophotoswipe.txt)
 * Single album
 * Album has 37 640x480 images

Performance is ... not good

````
37/37 [13:40<00:00, 22.17s/it]
````

Strace shows nothing usable. No verbose mode available so no idea what is happening on the inside.

````
$ pip freeze
Pillow==4.0.0
PyYAML==3.12
argparse==1.2.1
distribute==0.6.24
hugophotoswipe==0.0.12
olefile==0.44
pytz==2016.10
six==1.10.0
smartcrop==0.1
tqdm==4.11.2
tzlocal==1.3
wsgiref==0.1.2
````

**Comment by GjjvdBurg** [comment]
Wow that's some terribe performance! I've just uploaded a new version of HugoPhotoSwipe to PyPI which includes a verbose mode. Can you try again to debug where the biggest delay occurs? I suspect it's the SmartCrop functionality, in which case you can consider trying with SmartCrop.js to see if that speeds things up.

**Comment by zeridon** [comment]
Your suspicion is correct. The smartcrop functionality is the culprit here.

With the python implementation:
```
[2017-03-27 15:00:34] - [mavica] There are 37 photos to process.
[2017-03-27 15:00:34] - [MVC-039F.JPG] Creating large size.
[2017-03-27 15:00:34] - [MVC-039F.JPG] Not scaling up.
[2017-03-27 15:00:34] - [MVC-039F.JPG] Creating large image of dimensions: 640x480
[2017-03-27 15:00:34] - [MVC-039F.JPG] Saving large image to /home/zeridon/tmp/hps-test/output/mavica/large/mvc-039f_640x480.jpg
[2017-03-27 15:00:34] - [MVC-039F.JPG] Creating small size.
[2017-03-27 15:00:34] - [MVC-039F.JPG] Not scaling up.
[2017-03-27 15:00:34] - [MVC-039F.JPG] Creating small image of dimensions: 640x480
[2017-03-27 15:00:34] - [MVC-039F.JPG] Saving small image to /home/zeridon/tmp/hps-test/output/mavica/small/mvc-039f_640x480.jpg
[2017-03-27 15:00:34] - [MVC-039F.JPG] Creating thumbnail size.
[2017-03-27 15:00:34] - [MVC-039F.JPG] SmartCrop.py new dimensions: 256x256
[2017-03-27 15:00:34] - [MVC-039F.JPG] SmartCrop.py computing optimal crop size.
[2017-03-27 15:00:56] - [MVC-039F.JPG] Saving SmartCrop.py thumbnail.
```

With node implementation:

```
[2017-03-27 15:28:45] - [mavica] There are 37 photos to process.
[2017-03-27 15:28:45] - [MVC-039F.JPG] Creating large size.
[2017-03-27 15:28:45] - [MVC-039F.JPG] Not scaling up.
[2017-03-27 15:28:45] - [MVC-039F.JPG] Creating large image of dimensions: 640x480
[2017-03-27 15:28:45] - [MVC-039F.JPG] Saving large image to /home/zeridon/tmp/hps-test/output/mavica/large/mvc-039f_640x480.jpg
[2017-03-27 15:28:45] - [MVC-039F.JPG] Creating small size.
[2017-03-27 15:28:45] - [MVC-039F.JPG] Not scaling up.
[2017-03-27 15:28:45] - [MVC-039F.JPG] Creating small image of dimensions: 640x480
[2017-03-27 15:28:45] - [MVC-039F.JPG] Not scaling up.
[2017-03-27 15:28:45] - [MVC-039F.JPG] Saving small image to /home/zeridon/tmp/hps-test/output/mavica/small/mvc-039f_640x480.jpg
[2017-03-27 15:28:45] - [MVC-039F.JPG] Creating thumbnail size.
[2017-03-27 15:28:45] - [MVC-039F.JPG] SmartCrop.js new dimensions: 256x256
[2017-03-27 15:28:45] - [MVC-039F.JPG] SmartCrop.js running crop command.
```

I ran additional test. Using large images (that need to be downscaled) and there the python implementation is a bit better (about 4 sec/image)

```
[2017-03-27 14:59:27] - [d3200] There are 12 photos to process.
[2017-03-27 14:59:27] - [DSC_1769.JPG] Creating large size.
[2017-03-27 14:59:27] - [DSC_1769.JPG] Creating large image of dimensions: 1600x1063
[2017-03-27 14:59:28] - [DSC_1769.JPG] Saving large image to /home/zeridon/tmp/hps-test/output/d3200/large/dsc_1769_1600x1063.jpg
[2017-03-27 14:59:28] - [DSC_1769.JPG] Creating small size.
[2017-03-27 14:59:28] - [DSC_1769.JPG] Creating small image of dimensions: 800x531
[2017-03-27 14:59:28] - [DSC_1769.JPG] Saving small image to /home/zeridon/tmp/hps-test/output/d3200/small/dsc_1769_800x531.jpg
[2017-03-27 14:59:28] - [DSC_1769.JPG] Creating thumbnail size.
[2017-03-27 14:59:28] - [DSC_1769.JPG] SmartCrop.py new dimensions: 256x256
[2017-03-27 14:59:28] - [DSC_1769.JPG] SmartCrop.py computing optimal crop size.
[2017-03-27 14:59:32] - [DSC_1769.JPG] Saving SmartCrop.py thumbnail.
```

Using the node implementation on this dataset the speed is around 2 sec/image.

I can send you few sample shots (from the small ones) for testing.

**Comment by GjjvdBurg** [comment]
I've tried to reproduce this with some random 640x480 images, but I don't get the same slow performance. You could try to do the crop using the SmartCrop.py program directly (from the command line), to verify if it's SmartCrop.py or HugoPhotoSwipe. Please let me know what you find out, I'm curious to see where the problem lies.

For the record, I get about 4 sec/image using SmartCrop.py on large images as well.

**Comment by zeridon** [comment]
I've added a bit more debugging output and it is apparent that the smartcrop library calls are very slow

```
[2017-04-05 17:25:35] - [MVC-043F.JPG] Creating large size.
[2017-04-05 17:25:35] - [MVC-043F.JPG] Not scaling up.
[2017-04-05 17:25:35] - [MVC-043F.JPG] Creating large image of dimensions: 640x480
[2017-04-05 17:25:35] - [MVC-043F.JPG] Saving large image to /home/zeridon/tmp/hps-test/output/mavica/large/mvc-043f_640x480.jpg
[2017-04-05 17:25:35] - [MVC-043F.JPG] Creating small size.
[2017-04-05 17:25:35] - [MVC-043F.JPG] Not scaling up.
[2017-04-05 17:25:35] - [MVC-043F.JPG] Creating small image of dimensions: 640x480
[2017-04-05 17:25:35] - [MVC-043F.JPG] Saving small image to /home/zeridon/tmp/hps-test/output/mavica/small/mvc-043f_640x480.jpg
[2017-04-05 17:25:35] - [MVC-043F.JPG] Creating thumbnail size.
[2017-04-05 17:25:35] - [MVC-043F.JPG] SmartCrop.py new dimensions: 256x256
[2017-04-05 17:25:35] - [MVC-043F.JPG] SmartCrop.py computing optimal crop size.
[2017-04-05 17:25:35] - [MVC-043F.JPG] smartcrop -- start -- sc.crop(img, crop_options)
[2017-04-05 17:25:57] - [MVC-043F.JPG] smartcrop -- end   -- Crop complete
[2017-04-05 17:25:57] - [MVC-043F.JPG] nimg -- start -- self.original_image.crop(box)
[2017-04-05 17:25:57] - [MVC-043F.JPG] nimg -- end   -- self.original_image.crop(box)
[2017-04-05 17:25:57] - [MVC-043F.JPG] nimg -- start -- load()
[2017-04-05 17:25:57] - [MVC-043F.JPG] nimg -- end   -- load()
[2017-04-05 17:25:57] - [MVC-043F.JPG] nimg -- start -- nimg.thumbnail((nwidth, nheight), Image.ANTIALIAS)
[2017-04-05 17:25:57] - [MVC-043F.JPG] nimg -- end   -- nimg.thumbnail((nwidth, nheight), Image.ANTIALIAS)
[2017-04-05 17:25:57] - [MVC-043F.JPG] Saving SmartCrop.py thumbnail.
```

On the command line smartcroppy is really fast

```
$ time smartcroppy --debug --width 256 --height 256 MVC-039F.JPG MVC-039F-thumb.JPG
.....

real	0m0.515s
user	0m0.491s
sys	0m0.024s
```

looking further into smartcrop module

**Comment by zeridon** [comment]
After some more digging the slow code is in smartcrop.py

and i suspect it is this one
```
        for crop in crops:
            crop['score'] = self.score(score_output, crop)
            if crop['score']['total'] > topScore:
                topCrop = crop
                topScore = crop['score']['total']
```

Which maybe generates too many crops. If being called by the cmd line smartcroppy i get 77 possible crops and one top crop. When called like library they seem thousands.

Some debug extracts and the lines are in the thousands.

```
[2017-04-05 18:18:32] - [MVC-043F.JPG] smartcrop -- start -- sc.crop(img, crop_options)
[SmartCrop.crop] starting analyze
[SmartCrop.analyze] -- start -- detect_edge
[SmartCrop.analyze] -- end   -- detect_edge
[SmartCrop.analyze] -- start -- detect_skin
[SmartCrop.analyze] -- end   -- detect_skin
[SmartCrop.analyze] -- start -- detect_saturation
[SmartCrop.analyze] -- end   -- detect_saturation
[SmartCrop.analyze] -- start -- score_output = copy.copy
[SmartCrop.analyze] -- end   -- score_output = copy.copy
[SmartCrop.analyze] -- start -- score_output.thumbnail
[SmartCrop.analyze] -- end   -- score_output.thumbnail
[SmartCrop.crops] -- start -- scale
[SmartCrop.crops] -- end   -- scale
[SmartCrop.crops] -- start -- for scale in scales:
[SmartCrop.crops] -- end   -- for scale in scales:
[SmartCrop.analyze] -- start -- for crop in crops:
[SmartCrop.score] -- start scoring
[SmartCrop.score] -- start -- od = output.getdata()
[SmartCrop.score] -- end   -- od = output.getdata()
[SmartCrop.score] -- end scoring
[SmartCrop.score] -- start scoring
[SmartCrop.score] -- start -- od = output.getdata()
[SmartCrop.score] -- end   -- od = output.getdata()
[SmartCrop.score] -- end scoring
[SmartCrop.score] -- start scoring
[SmartCrop.score] -- start -- od = output.getdata()
[SmartCrop.score] -- end   -- od = output.getdata()
[SmartCrop.score] -- end scoring
....
[SmartCrop.score] -- start scoring
[SmartCrop.score] -- start -- od = output.getdata()
[SmartCrop.score] -- end   -- od = output.getdata()
[SmartCrop.score] -- end scoring
[SmartCrop.analyze] -- end   -- for crop in crops:
[SmartCrop.crop] finished analyze
[2017-04-05 18:24:18] - [MVC-043F.JPG] smartcrop -- end   -- Crop complete
[2017-04-05 18:24:18] - [MVC-043F.JPG] nimg -- start -- self.original_image.crop(box)
[2017-04-05 18:24:18] - [MVC-043F.JPG] nimg -- end   -- self.original_image.crop(box)
[2017-04-05 18:24:18] - [MVC-043F.JPG] nimg -- start -- load()
[2017-04-05 18:24:18] - [MVC-043F.JPG] nimg -- end   -- load()
[2017-04-05 18:24:18] - [MVC-043F.JPG] nimg -- start -- nimg.thumbnail((nwidth, nheight), Image.ANTIALIAS)
[2017-04-05 18:24:18] - [MVC-043F.JPG] nimg -- end   -- nimg.thumbnail((nwidth, nheight), Image.ANTIALIAS)
[2017-04-05 18:24:18] - [MVC-043F.JPG] Saving SmartCrop.py thumbnail
```

One possibility i see is somehow the scaling factors for smartcrop are set so absurdly small that it generates quite a few possible crops and then the scoring just eats time like hell.

**Comment by GjjvdBurg** [comment]
Thanks for the additional research @zeridon! Your'e completely right, the smartcrop.py executable generates way less crops than the HugoPhotoSwipe approach. SmartCrop.py always uses a width of 100 and scales the height accordingly,  whereas HPS uses the width and height of the desired output size. Since the desired output size in HPS is typically larger than 100x100, the HPS approach results in more crops (incidentally, if you use HPS with less than 100x100 as the desired output size, it'll be faster than Smartcrop). These additional crops indeed slow down HPS.

I'm not quite sure what the best solution is here, it's a trade-off between quality and speed. I guess I could add a flag to HPS for fast execution which would use the Smartcrop method. Could this help?

**Comment by GjjvdBurg** [comment]
Added fast mode with commit https://github.com/GjjvdBurg/HugoPhotoSwipe/commit/d4a342d052a99b69c4ddd718e3ee4064b8cca23f, available in version 0.0.14.
